vector/src/main/java/im/vector/app/features/onboarding/RegistrationWizardActionDelegate.kt

```diff
@@ -0,0 +1,123 @@
+            is RegisterAction.CaptchaDone                  -> resultOf { registrationWizard.performReCaptcha(action.captchaResponse) }
+            is RegisterAction.AcceptTerms                  -> resultOf { registrationWizard.acceptTerms() }
+            is RegisterAction.RegisterDummy                -> resultOf { registrationWizard.dummy() }
+            is RegisterAction.AddThreePid                  -> handleAddThreePid(registrationWizard, action)
+            is RegisterAction.SendAgainThreePid            -> resultOf { registrationWizard.sendAgainThreePid() }
+            is RegisterAction.ValidateThreePid             -> resultOf { registrationWizard.handleValidateThreePid(action.code) }
+            is RegisterAction.CheckIfEmailHasBeenValidated -> handleCheckIfEmailIsValidated(registrationWizard, action.delayMillis)
+            is RegisterAction.CreateAccount                -> resultOf {
+                registrationWizard.createAccount(
+                        action.username,
+                        action.password,
+                        action.initialDeviceName
+                )
+            }
+        }
+    }
+
+    private suspend fun handleAddThreePid(wizard: RegistrationWizard, action: RegisterAction.AddThreePid): RegistrationResult {
+        return runCatching { wizard.addThreePid(action.threePid) }.fold(
+                onSuccess = { it.toRegistrationResult() },
+                onFailure = {
+                    when {
+                        action.threePid is RegisterThreePid.Email && it.is401() -> RegistrationResult.SendEmailSuccess(action.threePid.email)
+                        else                                                    -> RegistrationResult.Error(it)
+                    }
+                }
+        )
+    }
+
+    private tailrec suspend fun handleCheckIfEmailIsValidated(registrationWizard: RegistrationWizard, delayMillis: Long): RegistrationResult {
+        return runCatching { registrationWizard.checkIfEmailHasBeenValidated(delayMillis) }.fold(
+                onSuccess = { it.toRegistrationResult() },
+                onFailure = {
+                    when {
+                        it.is401() -> null // recursively continue to check with a delay
+                        else       -> RegistrationResult.Error(it)
+                    }
+                }
+        ) ?: handleCheckIfEmailIsValidated(registrationWizard, 10_000)
+    }
+}
+
+private inline fun resultOf(block: () -> MatrixRegistrationResult): RegistrationResult {
+    return runCatching { block() }.fold(
+            onSuccess = { it.toRegistrationResult() },
+            onFailure = { RegistrationResult.Error(it) }
+    )
+}
+
+private fun MatrixRegistrationResult.toRegistrationResult() = when (this) {
+    is FlowResponse -> RegistrationResult.NextStep(flowResult)
+    is Success      -> RegistrationResult.Complete(session)
+}
+
+sealed interface RegistrationResult {
```

Comment:
I don't understand what is the role of this interface ? 
Which difference with:  RegistrationWizardAction.Result ? 
And why did you put Result in the Delegate class ?

Reply:
thanks for the explanation :)